cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -161,19 +192,210 @@ const requestCertificate = async function (
+            domainName: domainName,
+          })
+        );
+        break;
+    }
+  }
+  return Promise.all(promises);
+};
+
+const deleteHostedZoneRecords = async function (
+  options,
+  envName,
+  appName,
+  domainName,
+  envRoute53,
+  appRoute53,
+  acm,
+  envHostedZoneId
+) {
+  let isLegacyCert = false;
+  // Legacy cert only has two DomainValidationOptions:
+  // `${envName}.${appName}.${domainName}` and `*.${envName}.${appName}.${domainName}`
+  if (options.length <= 2) {
+    isLegacyCert = true;
+  }
+
+  let certWithDomainCount = await numOfGeneratedCertificates(
+    acm,
+    `${envName}.${appName}.${domainName}`
+  );
+  let isLastOne = false;
+  if (certWithDomainCount < 2) {
+    isLastOne = true;
+  }
+
+  const newOptions = [];
+  switch (`${isLegacyCert}|${isLastOne}`) {
+    case `true|true`:
+      // If it is a legacy cert and it is the last Copilot cert,
+      // we'll go ahead to remove the validation CNAME record in env hosted zone.
+      newOptions.push(...options);
+      break;
+    case `true|false`:
+      // If it is a legacy cert but it is not the last Copilot cert,
+      // we'll do nothing since the new Copilot cert needs validation CNAME record
+      // in the env hosted zone.
+      break;
+    case `false|true`:
+      // If it is not a legacy cert and it is the last Copilot cert,
+      // we'll remove all validation CNAME records in env/app/root hosted zone.
+      newOptions.push(...options);
+      break;
+    case `false|false`:
+      // If it is not a legacy cert and it is not the last Copilot cert,
+      // we'll remove validation CNAME records only for app and root hosted zone,
+      // since the legacy cert still needs the validation record in the env hosted zone.
+      for (const option of options) {
+        if (option.DomainName === `${envName}.${appName}.${domainName}`) {
+          continue;
+        }
+        newOptions.push(option);
+      }
+      break;
+  }
+  await updateHostedZoneRecords(
+    "DELETE",
+    newOptions,
+    envName,
+    appName,
+    domainName,
+    envRoute53,
+    appRoute53,
+    envHostedZoneId
+  );
+};
+
+// numOfGeneratedCertificates returns the number of Copilot generated certificates. Specifically it returns
+// the number of certificate with domain name as `${envName}.${appName}.${domainName}`, since if any of
+// the other certificate with this domain name still exists, we cannot remove the validation record in our
+// env hosted zone.
+const numOfGeneratedCertificates = async function (acm, defaultEnvDomain) {
+  let certWithDomainCount = 0;
+  let listCertResp = await acm.listCertificates({}).promise();
+  for (const certSummary of listCertResp.CertificateSummaryList || []) {
+    if (certSummary.DomainName === defaultEnvDomain) {
+      certWithDomainCount++;
+    }
+  }
+  let nextToken = listCertResp.NextToken;
+  // Though very unlikely, we'll handle the pagination.
+  while (nextToken && certWithDomainCount < 2) {
```

Comment:
lol i added `certWithDomainCount < 2` because i was worried it would take too much time and wanted to exit earlier. But yeah let's make the caller to handle that.

Reply:
Yeah that totally makes sense 😄 , in that scenario we can make it explicit with the function signature and a default parameter. For example:
```javascript
const numOfGeneratedCertificates = async function (acm, domainName, maxCount = 2)
```

Reply:
ok done